### ISSUE TITLE
More selectors

### ISSUE THREAD

**Issue by mrchantey** [mrchantey/beet#3]
Bonsai has lots of cool internal node options [here](https://github.com/Sollimann/bonsai/blob/9fb29e64646717b8a64018f3afd9750af656f208/bonsai/src/behavior.rs#L6), maybe we should implement some more in beet.

**Comment by mgi388** [comment]
`While` would be great to have. https://github.com/Sollimann/bonsai/blob/ec272b35feb83275b29188017d4af946d71d3d4a/examples/src/async_drone/main.rs#L207-L211

```rust
let behavior = While(
    Box::new(avoid_others),
    // takeoff, do mission, land again
    vec![takeoff, mission_with_fallback, land],
);
```

Being able to continually check a condition would make this really powerful. I don't think this particular type of node/action is possible at the moment with what you have? In my case, I just want to run the children/child of a `While` node if I'm in a certain state.

Another common example would be distance-based checks like this from bonsai. https://github.com/Sollimann/bonsai/blob/835f528a9a7895341e5efa9e22d3ccadce137618/examples/src/graphviz/main.rs#L33-L36

```rust
While(
    Box::new(Action(AttackDrone::PlayerWithinDistance(50.0))),
    vec![circling],
),
```

It would be fun to add to the seek_3d example to do something like wrapping the main BT in a `While` which only runs when it's (imaginary) dawn or dusk.

**Comment by mrchantey** [comment]
> While would be great to have.
> ```rust 
> let behavior = While(
>    Box::new(avoid_others),
>    vec![takeoff, mission_with_fallback, land],
>);
> ```
>

Very interesting use case! I hadn't yet come across a nested sequence to be interrupted like this and it led to an improvement of how bubbling for `RepeatFlow` works in beet.

In Beet the above could be achieved like this:

```rust
app.world_mut()
  .spawn((
    Name::new("Do Missions while out of enemy proximity"), 
    SequenceFlow
    // this doesnt yet exist but would check agent's proximity to any transform with `IsEnemy`
    // and trigger OnRunResult when closer than 0.5 meters
    EndInProximity::<IsEnemy>::new(0.5),
  )).with_children(|parent| {
    parent.spawn((
        Name::new("Mission Sequence"),
        SequenceFlow,
        // its important that the RepeatFlow is a child of the EndInProximity
        // otherwise it would just kick the root sequence off again
        RepeatFlow::default(),
    ))
      .with_child(SomeTakeOffAction)
      .with_child(SomeMissionAction)
      .with_child(SomeLandAction)
  });
```

A couple of extra notes:
- ⚠️ This will only work on main until `0.0.5` because of a change in how RepeatFlow bubbles results
- The reason `EndInProximity` stops the child sequence is because in beet interrupts are enabled by default, ie whenever `OnRunResult` is called on a parent, the `Running` component is removed recursively from all children.
- Requiring users to nest the repeat feels like extra work, as you mentioned at some point it may still be worth creating a dedicated `WhileFlow` to reduce complexity.
- The closest thing i have to this kind of proximity action is [`EndOnArrive`](https://github.com/mrchantey/beet/blob/46372e77ee31502cd6a1583af5725ce105dddf62/crates/beet_spatial/src/steer/steer_actions/end_on_arrive.rs) so maybe that can help for building something similar.
- See [`examples/long_running.rs`](https://github.com/mrchantey/beet/blob/main/examples/long_running.rs) for a full working example.

**Comment by mrchantey** [comment]
Ok I've created a proper behavior tree `While` pattern. the above behaves more like an `Interrupt` that checks the predicate regardless of the state of the repetition, whereas a `While` should only check the predicate at the beginning or end of each repetition.

Please see [`examples/repeat_while.rs`](https://github.com/mrchantey/beet/blob/main/examples/repeat_while.rs) for a full example.

The core idea is to use a `RunResult::Failure` to exit the repeat, this pattern also has the advantage that the actual sequence can also cancel the repeat.

```rust
app.world_mut()
  .spawn((
    SequenceFlow,
    // will repeat if the sequence succeeds, otherwise bubble up failure
    RepeatFlow::if_success()
  ))
  // replace this with a proximity check that fails if the enemy is near
  .with_child(SucceedTimes::new(2))
  // replace these with some actions to repeat
  .with_child(EndOnRun::success())
  .with_child(EndOnRun::success())
```

**Comment by mgi388** [comment]
I think the behavior shown in the example below is what I was mostly trying to construct by looking at a "while" flow in beet and bonsai.

The flow continually re-evaluates the `State` resource and if it's not "on", then the flow stops. In my game, I am simply trying to avoid running my enemy unit AI until the battle has started. Even though the example below works, I'd probably hoist it up into another root `Sequence` so that the `SucceedIfStateIsOn` behavior is shared by all child-trees of that root.

One thing that stands out to me as possibly confusing is that I can write the example below and it works like an infinitely running while loop (not infinite in the sense of a program crash). But, it's doing it using a `Sequence` as the root. From looking at other BT examples across other libraries and game engines, their sequence nodes were just control nodes and there was some other node/means to make the tree loop indefinitely (citation required!). In Beet, and the example below, it's the _combo_ of `SequenceFlow` _and_ `RepeatFlow` that's making it repeat / making it a while loop. It seems powerful to have it as two components, but it could be confusing. It's also possibly just me, so don't feel the need to change it from this feedback.

```rust
use beet::prelude::*;
use bevy::prelude::*;

fn main() {
	let mut app = App::new();
	app.add_plugins((
		DefaultPlugins,
		LifecyclePlugin,
		BeetDebugPlugin,
		ActionPlugin::<SucceedIfStateIsOn>::default(),
	))
	.insert_resource(State::default())
	.add_systems(Update, toggle_state)
	.world_mut()
	.spawn((Name::new("root"), SequenceFlow, RepeatFlow::default()))
	.observe(|_: Trigger<OnRun>| {
		println!("root triggered");
	})
	.with_child((Name::new("check state is on"), SucceedIfStateIsOn))
	.with_child((Name::new("fails on third run"), SucceedTimes::new(2)))
	.with_child((Name::new("some action to perform"), EndOnRun::success()))
	.trigger(OnRun);

	app.run();
}

#[derive(Resource, Default)]
struct State {
	on: bool,
}

#[derive(Debug, Default, Clone, PartialEq, Component, Reflect, Action)]
#[observers(succeed_if_in_state)]
#[reflect(Default, Component)]
struct SucceedIfStateIsOn;

fn succeed_if_in_state(
	trigger: Trigger<OnRun>,
	mut commands: Commands,
	state: Res<State>,
) {
	if state.on {
		println!("state is on, succeeding");
		commands
			.entity(trigger.entity())
			.trigger(OnRunResult::success());
	} else {
		println!("state is off, failing");
		commands
			.entity(trigger.entity())
			.trigger(OnRunResult::failure());
	}
}

fn toggle_state(input: Res<ButtonInput<KeyCode>>, mut state: ResMut<State>) {
	if input.just_pressed(KeyCode::Space) {
		state.on = !state.on;
		println!("state is now {:?}", state.on);
	}
}

```

**Comment by mrchantey** [comment]
Yeah i totally understand how that can be confusing, ECS does open up a whole bunch of patterns and antipatterns and we probably just need to experiment a lot to work out which is which.

I am looking forward to things like the Spawn Api you mentioned, it may make composing these kinds of structures more intuitive.

**Comment by mrchantey** [comment]
I'm going to close this issue because I've phrased it with an infinite scope, lets create other smaller issues.
